middleware/reverse/network.go

```diff
@@ -0,0 +1,133 @@
+package reverse
+
+import (
+	"net"
+	"regexp"
+	"bytes"
+	"strings"
+)
+
+type Network struct {
+	Ipnet    *net.IPNet
+	Domain   string
+	Template string
+	Ttl      uint32
+}
+
+const hexDigit = "0123456789abcdef"
+const templateNameIp = "{ip}"
+const regexMatchV4 = "((?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\-){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))"
+const regexMatchV6 = "([0-9a-fA-F]{32})"
+
+// Forward lookup
+func (network *Network) hostnameToIp(rname string) net.IP {
+	var matchedIp net.IP
+
+	if network.Ipnet.IP.To4() != nil {
+		// search for an v4 ip
+		r, _ := regexp.Compile(
+			strings.Replace(
+				network.Template,
+				templateNameIp,
+				regexMatchV4,
+				1,
+			))
+		match := r.FindStringSubmatch(rname)
+		if (len(match) == 2) {
+			matchedIp = net.ParseIP(strings.Replace(match[1], "-", ".", 4))
+		}
+	} else {
+		// search for an v6 ip
+		r, _ := regexp.Compile(
+			strings.Replace(
+				network.Template,
+				templateNameIp,
+				regexMatchV6,
+				1,
+			))
+		match := r.FindStringSubmatch(rname)
+		if (len(match) == 2) {
+
+			var buf bytes.Buffer
+			for i := 0; i < 8 * 4; i += 4 {
+				buf.WriteString(match[1][i:i + 4])
+				if (i < 28) {
+					buf.WriteString(":")
+				}
+			}
+			matchedIp = net.ParseIP(buf.String())
+		}
+	}
+
+	if matchedIp == nil || !network.Ipnet.Contains(matchedIp) {
+		return nil
+	}
+
+	return matchedIp
+}
+
+// Reverse lookup
+func (network *Network) ipToHostname(ip *net.IP) string {
```

Comment:
`net.IP` is an interface type, no? Then drop the pointer, there is no need for it.